Eye Dome Lighting for point clouds in 2D map canvas

NEDJIMAbelgacem opened the pull request:
## Description
This PR implements eye dome lighting effect for point clouds 2D renderers.
For now it only works well when the sorted rendering is applied,
### Eye dome lighting applied to a renderer using a single color:
![Screenshot from 2022-07-19 16-14-21](https://user-images.githubusercontent.com/29183781/179786628-006713b7-ff01-4a40-b436-26c3b5757751.png)
### Trencin castle dataset without eye dome lighting applied:
![Screenshot from 2022-07-19 16-15-21](https://user-images.githubusercontent.com/29183781/179786835-615f1011-1abc-4cfb-bfde-7cee4e6218b4.png)
### Trencin castle dataset with eye dome lighting applied:
![Screenshot from 2022-07-19 16-14-53](https://user-images.githubusercontent.com/29183781/179787028-44057550-75bb-45c5-874d-c080d0b7886a.png)

### GUI

![image](https://user-images.githubusercontent.com/193367/187996287-5e7f6a53-fb0b-4437-b039-8e74023371ec.png)

NEDJIMAbelgacem committed:
initial implementation

wonder-sk commented:
This is an initial attempt to get eye dome lighting working in 2D canvas for better depth perception. Currently the implementation is completely tied to point cloud layers, where the 2D renderer internally builds a depth map and then applies the shading as a post-processing effect.

There are two major directions how this could be further improved:
1. support other kinds of post-processing of depth map - we could have hillshading as well (by taking the implementation from the raster pipeline) or other effects (contours, ambient occlusion, ...)
2. support other map layers - extend this functionality to take into account all map layers with 3D data (rasters, vectors, meshes), compose depth maps of individual map layers and do the post-processing effect on the final depth map.

Especially the second option seems like quite a lot of extra work, so we'll probably stick to the per-point-cloud-layer implementation for now and revisit this when there's budget for more work on this...

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -67,12 +67,78 @@ QgsPointCloudLayerRenderer::QgsPointCloudLayerRenderer( QgsPointCloudLayer *laye
+  {
+    for ( int j = distance; j < img->height() - distance; ++j )
+    {
+      QColor originalColor = img->pixelColor( i, j );
```
Calling pixelColor for each pixel is extremely inefficient. Best to read the image data directly using scan lines or QImage::constBits. See https://doc.qt.io/qt-5/qimage.html#pixelColor

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -67,12 +67,78 @@ QgsPointCloudLayerRenderer::QgsPointCloudLayerRenderer( QgsPointCloudLayer *laye
+      if ( sampledPointsCount == 0 )
+        continue;
+      float shade = exp( -factor / sampledPointsCount * strength );
+      img->setPixelColor( i, j, QColor::fromRgbF( shade * originalColor.redF(), shade * originalColor.greenF(), shade * originalColor.blueF(), originalColor.alphaF() ) );
```
Same here -- setPixelColor is very slow for bulk operations. Best to manipulate the QImage data direct. https://doc.qt.io/qt-5/qimage.html#pixelColor

NEDJIMAbelgacem replied:
But I believe if I manipulate QImage directly I should assume one format of QImage being used which seems like bad idea if the format of the rendered image is changed in the future.

nyalldawson commented on src/core/pointcloud/qgspointcloudlayerrenderer.cpp:
```diff
@@ -67,12 +67,78 @@ QgsPointCloudLayerRenderer::QgsPointCloudLayerRenderer( QgsPointCloudLayer *laye
 
   // Set up the render configuration options
   QPainter *painter = context.renderContext().painter();
+  std::unique_ptr<QImage> elevationImage;
```
This could be stack allocated

NEDJIMAbelgacem replied:
Because it is optional I made it into a unique pointer, I prefer doing it this way because that allocating a big image on the stack is not a very good idea in my opinion.

nyalldawson commented on python/core/auto_generated/pointcloud/qgspointcloudrenderer.sip.in:
```diff
@@ -48,6 +48,78 @@ Returns a reference to the context's render context.
 
 
+
+    void setElevationPainter( QPainter *painter );
```
I'd SIP_SKIP all these -- I don't think they should be part of stable API.

Eg. the elevation painter would need to move to QgsRenderContext for use outside of point clouds.

nyalldawson commented on src/ui/pointcloud/qgspointcloudrendererpropsdialogbase.ui:
```diff
@@ -233,20 +229,80 @@
+                </widget>
+               </item>
+               <item row="1" column="1">
+                <widget class="QSpinBox" name="mEdlStrength">
```
Make these QgsSpinBox with default clear values set.

nyalldawson commented on src/ui/pointcloud/qgspointcloudrendererpropsdialogbase.ui:
```diff
@@ -233,20 +229,80 @@
+                </widget>
+               </item>
+               <item row="0" column="0" colspan="2">
+                <widget class="QCheckBox" name="mApplyEdlCheckbox">
```
You can avoid the extra checkbox widget by making the Eye Dome groupbox checkable instead

nyalldawson commented:
@NEDJIMAbelgacem 
> For now it only works well when the sorted rendering is applied,

I think there's a way to avoid this -- if you just wrote grayscale colors to the elevation painter, you could set the painter composition mode to either lighten or darken so that it doesn't matter what order the points are rendered, you'll always get the top most elevation rendered.

If needed, you could change the elevation image to QImage::Format_Grayscale16 and use grayscale colors created using the float QColor methods to retain the higher precision elevation values.

uclaros reviewed (commented):
Nice work!
It would be also nice to state at the gui that edl is only usable with ordered rendering

nyalldawson commented:
@uclaros 

>  would be also nice to state at the gui that edl is only usable with ordered rendering

We should be able to avoid that limitation entirely -- see my comment aboce

uclaros commented:
@nyalldawson 
> We should be able to avoid that limitation entirely -- see my comment above

I've seen your comment and that's a clever approach! I'm not sure if the result will actually make sense when applied to an unordered point cloud (maybe it will after all, since not all point clouds have occluded points)

We could also re-implement ordered rendering to use this technique:
- lighten/darken the depth map
- check if depth map has changed
- skip drawing the actual point if depth map hasn't changed
We could then set ordered rendering as default, since it will now not require all points to be fetched and ordered first!

Major drawback: `QImage::Format_Grayscale16` seems to be introduced in qt 5.13

nyalldawson commented:
> seems to be introduced in qt 5.13

Isn't that the current min anyway? (It should really be 5.15 now!)

NEDJIMAbelgacem committed:
switch to using grayscale image and fix the default draw order rendering

NEDJIMAbelgacem committed:
address reviews

NEDJIMAbelgacem committed:
Revert to RGBA elevation

NEDJIMAbelgacem commented:
> @NEDJIMAbelgacem
> 
> > For now it only works well when the sorted rendering is applied,
> 
> I think there's a way to avoid this -- if you just wrote grayscale colors to the elevation painter, you could set the painter composition mode to either lighten or darken so that it doesn't matter what order the points are rendered, you'll always get the top most elevation rendered.
> 
> If needed, you could change the elevation image to QImage::Format_Grayscale16 and use grayscale colors created using the float QColor methods to retain the higher precision elevation values.

Actually I managed to make the EDL work with unordered rendering without the need for a QImage::Format_Grayscale16 image. The actual order of elevation values in the elevation image doesn't actually matter in our case because the colors themselves are not sorted by elevation.

NEDJIMAbelgacem committed:
address reviews and fix docs

NEDJIMAbelgacem committed:
fix layout tests

nyalldawson commented on src/core/pointcloud/qgspointcloudrenderer.h:
```diff
@@ -622,6 +749,43 @@ class CORE_EXPORT QgsPointCloudRenderer
+      const double zMax = +5000;//context.zMax();// 532.87;
+      double zFloat = ( z - zMin ) / ( zMax - zMin );
+      context.updateZRange( zFloat );
+      zFloat = std::max<double>( 0.0, std::min<double>( 1.0, zFloat ) );
```
We already require c++17 -- clamp is used a lot throughout the code

NEDJIMAbelgacem committed:
use std::clamp and bit manupilation for elevation colors

wonder-sk commented on src/core/pointcloud/qgspointcloudrenderer.cpp:
```diff
@@ -159,6 +159,43 @@ QStringList QgsPointCloudRenderer::legendRuleKeys() const
+  const double zMax = +5000;
+  float zFloat = ( z - zMin ) / ( zMax - zMin );
+  zFloat = std::clamp<float>( zFloat, 0.0, 1.0 );
+  context.updateZRange( zFloat );
```
I'm confused... here the depth values are transformed from [-5000, +5000] to [0,1] but then in applyEyeDomeLighting() they are interpreted as being in [zMin,zMax] range as stored in the context, based on actual values?

wonder-sk reviewed (commented):
The output is not looking great compared to the 3D view's EDL - probably a bug?

![image (14)](https://user-images.githubusercontent.com/193367/182119291-0306297a-6f31-404f-a893-6608584b5b59.png)

NEDJIMAbelgacem committed:
use image bits for writing

wonder-sk committed:
Eye dome lighting code fixes

- make it independent on the current view's Z range
- simplify scaling code
- reverse Z handling (higher Z value == closer to "camera")
- remove unused API functions and member variables

wonder-sk committed:
Refactor elevation map related code to a new class QgsElevationMap

wonder-sk committed:
Remove refactored code

wonder-sk committed:
Make 2px EDL distance default (same as in 3D)

wonder-sk committed:
Fix sip

wonder-sk commented:
I have fixed a couple of rendering bugs so the results of 2D and 3D shading are quite comparable:

![image](https://user-images.githubusercontent.com/193367/184246263-7192facc-dab6-4020-8eed-5d55f740dd10.png)

Also refactored the code so that most of the code resides in a new `QgsElevationMap` class which could be later used also for other kinds of shading based on depths, and it could provide functionality like combination of elevation maps from multiple sources to a single one.

To-do:
- depth encoding could be improved / cleaned up
- how to deal with the case where "normal" drawing order is used (shading may not be correct)

wonder-sk committed:
Better encoding of elevation values

wonder-sk committed:
Disable copying of QgsElevationMap

wonder-sk committed:
Add a warning to EDL when draw order is default

wonder-sk commented:
There is a warning now if EDL is being used with the default draw order (not shown when some other draw order mode is used)

![image](https://user-images.githubusercontent.com/193367/184378367-fb25f6a5-8ce1-4f0d-b3ce-641b602bd40a.png)

wonder-sk committed:
Add tests for QgsElevationMap and for EDL of 2D point cloud rendering

wonder-sk commented:
Everything that was on my to-do list is now addressed, the PR should be ready for a final review...

nyalldawson commented:
(triggering a rebuild to run through new qt 6 workflows...)

uclaros reviewed (commented):
The effect looks great!
I've added some minor suggestions.

nyalldawson commented on python/core/auto_generated/pointcloud/qgspointcloudrenderer.sip.in:
```diff
@@ -156,6 +156,24 @@ Returns the feedback object used to cancel rendering
 .. versionadded:: 3.20
 %End
 
+    void setElevationMap( QgsElevationMap *elevationMap /Transfer/ );
```
How do you feel about hiding these methods from sip? There's a good chance we may want to shuffle them around in future (eg moving them to QgsRenderContext), and I wouldn't want to be constrained by stable API issues if/when we do this.

nyalldawson commented on src/core/pointcloud/qgspointcloudrenderer.h:
```diff
@@ -570,6 +587,42 @@ class CORE_EXPORT QgsPointCloudRenderer
+     * Returns the eye dome lighting distance value
+     * \since QGIS 3.28
+     */
+    int eyeDomeLightingDistance() const { return mEyeDomeLightingDistance; }
```
What's the units here? It would be better to make this a double and add a corresponding QgsUnitTypes::RenderUnit parameter, and then convert to painter units using that. Otherwise we won't get consistent results from canvas vs hi resolution layout exports.

wonder-sk replied:
it is in pixels (will add that). Using floating point numbers for pixel distance probably wouldn't make sense - interpolating elevation values of neighboring pixels seems wrong (because of discontinuities).

I am wondering if it would make sense to have the distance e.g. in millimeters, but then we would convert it to integer pixel value - it may be weird that small adjustments of the value wouldn't trigger any change in behavior...

nyalldawson replied:
That's why I'd go the double + unit route. It's what we do elsewhere to handle this, even for situations where the final number must be an integer pixel size (eg font sizes, blur radiuses, etc). Otherwise it'll be impossible to use this effect on print layouts, since the pixel sizes for screen won't be visible at print resolution outputs. (Also there'll be issues sharing projects across hi DPI/non hi DPI systems)

nyalldawson commented on src/core/qgselevationmap.cpp:
```diff
@@ -0,0 +1,90 @@
+{
+  mElevationImage.fill( 0 );
+  mPainter.reset( new QPainter );
+  mPainter->begin( &mElevationImage );
```
Won't destroying a painter in an active state cause issues? (via the default destructor)

wonder-sk replied:
It seems that should be fine - if you are referring to a missing QPainter::end() call - it is optional according to the docs https://doc.qt.io/qt-5/qpainter.html#end

nyalldawson replied:
Hmm that's odd! I often get crashes from python code when painters are destroyed while still in use.. maybe that's a PyQt "thing" though 🤨

nyalldawson commented on src/core/qgselevationmap.cpp:
```diff
@@ -0,0 +1,90 @@
+    for ( int j = distance; j < img.height() - distance; ++j )
+    {
+      float factor = 0.0f;
+      float centerDepth = decodeElevation( elevPtr[ j * imgWidth + i ] );
```
There's potential for an overflow here:

```suggestion
      float centerDepth = decodeElevation( elevPtr[ j * static_cast< qgsSize >( imgWidth ) + i ] );
```

nyalldawson commented on src/core/qgselevationmap.cpp:
```diff
@@ -0,0 +1,90 @@
+      {
+        int iNeighbour = i + distance * neighbours[2 * k];
+        int jNeighbour = j + distance * neighbours[2 * k + 1];
+        float neighbourDepth = decodeElevation( elevPtr[ jNeighbour * imgWidth + iNeighbour ] );
```
```suggestion
        float neighbourDepth = decodeElevation( elevPtr[ jNeighbour * static_cast< qgsSize >( imgWidth ) + iNeighbour ] );
```

nyalldawson commented on src/core/qgselevationmap.cpp:
```diff
@@ -0,0 +1,90 @@
+      }
+      float shade = exp( -factor * strength / rendererScale );
+
+      QRgb c = imgPtr[ j * imgWidth + i ];
```
```suggestion
      QRgb c = imgPtr[ j * static_cast< qgsSize >( imgWidth ) + i ];
```

nyalldawson commented on src/core/qgselevationmap.cpp:
```diff
@@ -0,0 +1,90 @@
+      float shade = exp( -factor * strength / rendererScale );
+
+      QRgb c = imgPtr[ j * imgWidth + i ];
+      imgPtr[ j * imgWidth + i ] = qRgba( qRed( c ) * shade, qGreen( c ) * shade, qBlue( c ) * shade, qAlpha( c ) );
```
```suggestion
      imgPtr[ j * static_cast< qgsSize >( imgWidth ) + i ] = qRgba( qRed( c ) * shade, qGreen( c ) * shade, qBlue( c ) * shade, qAlpha( c ) );
```

That said, there's a lot of imgPtr[ ...] happening here. Could we store it once in a pointer upfront and avoid the multiple [.... ] calls?

nyalldawson commented on tests/src/core/testqgselevationmap.cpp:
```diff
@@ -0,0 +1,107 @@
+  return QDir::tempPath() + '/' + testName + ".png";
+}
+
+static bool _verifyImage( const QString &testName, QString &report )
```
Can you use the base class method QgsTest::imageCheck here instead? I'm trying to consolidate all the duplicate render checking code into that base class

wonder-sk committed:
Address reviews from Stefanos and Nyall

wonder-sk committed:
Fix build and switch to QgsTest::imageCheck()

wonder-sk committed:
Allow any render unit for EDL distance as suggested by Nyall

This will make it possible to get equivalent effect also on different DPI

wonder-sk committed:
Bring back some Python methods to make the test in Python work again

wonder-sk committed:
fix function name in test
